include/jemalloc/internal/psset.h

```diff
@@ -0,0 +1,61 @@
+#ifndef JEMALLOC_INTERNAL_PSSET_H
+#define JEMALLOC_INTERNAL_PSSET_H
+
+/*
+ * A page-slab set.  What the eset is to PAC, the psset is to HPA.  It maintains
+ * a collection of page-slabs (the intent being that they are backed by
+ * hugepages, or at least could be), and handles allocation and deallocation
+ * requests.
+ *
+ * It has the same synchronization guarantees as the eset; stats queries don't
+ * need any external synchronization, everything else does.
+ */
+
+/*
+ * One more than the maximum pszind_t we will serve out of the HPA.
+ * Practically, we expect only the first few to be actually used.  This
+ * corresponds to a maximum size of of 512MB on systems with 4k pages and
+ * SC_NGROUP == 4, which is already an unreasonably large maximum.  Morally, you
+ * can think of this as being SC_NPSIZES, but there's no sense in wasting that
+ * much space in the arena, making bitmaps that much larger, etc.
+ */
+#define PSSET_NPSIZES 64
+
+typedef struct psset_s psset_t;
+struct psset_s {
+	/*
+	 * The pageslabs, quantized by the size class of the largest contiguous
+	 * free run of pages in a pageslab.
+	 */
+	edata_heap_t pageslabs[PSSET_NPSIZES];
+	bitmap_t bitmap[BITMAP_GROUPS(PSSET_NPSIZES)];
+};
+
+void psset_init(psset_t *psset);
+
+
+/*
+ * Tries to obtain a chunk from an existing pageslab already in the set.
+ * Returns true on failure.
+ */
+bool psset_alloc_reuse(psset_t *psset, edata_t *r_edata, size_t size);
```

Comment:
I.e. and also have the psset do the alloc itself (calling into the hpa_central_t)?

I think testability, but also just like various other design considerations get easier if this is a "pure" container (in the sense of not affecting any mappings / OS-level state, analogous to the eset). It doesn't have to do locking, know about VM management, etc. In the commits that tie everything together, we do something like the grow-retained trick in the ecache. This split lets us have one portion of code that purely does extent management, and another bit of code that purely handles locking trickery, rather than mixing them.

(Those commits are at https://github.com/davidtgoldblatt/jemalloc/tree/HPA-merged, but there's not a pull request out for them yet; see in particular hpa.c in https://github.com/davidtgoldblatt/jemalloc/commit/a80c97b67fde9e8ff8833c2ee479f4be6eb2e70c ).

Reply:
Yeah I prefer the current split API as well. I was more asking for my curiosity / trying to understand the use cases.